docs/knowledgebase/runtime/benchmarking.md

```diff
@@ -0,0 +1,561 @@
+
+```bash
+cargo build --release --features runtime-benchmarks
+```
+
+Once this is done, you should be able to run the `benchmark` subcommand.
+
+```bash
+cd ../../../
+target/release/substrate benchmark --help
+```
+
+The Benchmarking CLI has a lot of options which can help you automate your benchmarking. A minimal
+command to execute benchmarks for a pallet will look like this:
+
+```bash
+./target/release/substrate benchmark \
+    --chain dev \               # Configurable Chain Spec
+    --execution wasm \          # Always test with Wasm
+    --wasm-execution compiled \ # Always used `wasm-time`
+    --pallet pallet_example \   # Select the pallet
+    --extrinsic '\*' \          # Select the benchmark case name, using '*' for all
+    --steps 20 \                # Number of steps across component ranges
+    --repeat 10 \               # Number of times we repeat a benchmark
+    --raw \                     # Optionally output raw benchmark data to stdout
+    --output ./                 # Output results into a Rust file
+```
+
+- `chain` specifies the chain-spec we will use
+- `execution` and `wasm-execution` specify we are benchmarking the compiled wasm code (vs the native
+  code as an alternative)
+- `pallet` and `extrinsic` specify which pallet and extrinsics we are benchmarking. With the above
+  arguments we are benchmarking all extrinsics of `pallet_example`.
+- `steps` and `repeat` means how many steps will be taken to walk through each variable range, and
+  how many time the execution state will be repeated.
+
+Let's take `accumulate_dummy` benchmark case as an example.
+
+```rust
+accumulate_dummy {
+  let b in 1 .. 1000;
+  let caller = account("caller", 0, 0);
+}: _ (RawOrigin::Signed(caller), b.into())
+```
+
+With `--steps 20 --repeat 10` in the benchmark input arguments, `b` will walk 20 steps to reach
+1,000, so `b` will start from 1 and increment by about 50. For each value of `b`, we will execute
+the benchmark 10 times and record the benchmark information.
+
+![multi-variables-regression](assets/runtime/benchmarking/multi-variables-regression.png)
+
+A more complete command that we use for the Substrate benchmarking output looks like this:
+
+```bash
+cargo run --release
+  --features=runtime-benchmarks
+  --manifest-path=bin/node/cli/Cargo.toml
+  --
+  benchmark
+  --chain=dev
+  --steps=50
+  --repeat=20
+  --pallet=frame_system
+  --extrinsic="*"
+  --execution=wasm
+  --wasm-execution=compiled
+  --heap-pages=4096
+  --output=./frame/system/src/weights.rs
+  --template=./.maintain/frame-weight-template.hbs
+```
+
+However, the exact command you run will depend on your needs.
+
+> **Notes**
+>
+> Another argument introduced above is `--template`. We allow you to specify your own weight
+> template file and the benchmarking toolchain will fill in the exact numbers from the measured
+> result. This enables automating weight generation in your desired code format and integrates this
+> in your CI process.
+
```

Comment:
@kianenigma added a paragraph here telling readers they can use `--template` to customize the weightinfo implementation.